src/transformation/visitors/optional-chaining.ts

```diff
@@ -0,0 +1,187 @@
+import * as ts from "typescript";
+import * as lua from "../../LuaAST";
+import { TransformationContext } from "../context";
+import { assert, assertNever } from "../../utils";
+import { transformInPrecedingStatementScope } from "../utils/preceding-statements";
+import { transformPropertyAccessExpressionWithCapture, transformElementAccessExpressionWithCapture } from "./access";
+import { shouldMoveToTemp } from "./expression-list";
+
+function flattenChain(chain: ts.OptionalChain) {
+    assert(!ts.isNonNullChain(chain));
+    const links: ts.OptionalChain[] = [chain];
+    while (!chain.questionDotToken && !ts.isTaggedTemplateExpression(chain)) {
+        const nextLink: ts.Expression = chain.expression;
+        assert(ts.isOptionalChain(nextLink) && !ts.isNonNullChain(nextLink));
+        chain = nextLink;
+        links.unshift(chain);
+    }
+    return { expression: chain.expression, chain: links };
+}
+
+export interface ExpressionWithThisValue {
+    expression: lua.Expression;
+    thisValue?: lua.Expression;
+    isNewTemp?: boolean;
+}
+
+function transformExpressionWithThisValueCapture(
+    context: TransformationContext,
+    node: ts.Expression
+): ExpressionWithThisValue {
+    if (ts.isParenthesizedExpression(node)) {
+        return transformExpressionWithThisValueCapture(context, node.expression);
+    }
+    if (ts.isPropertyAccessExpression(node)) {
+        return transformPropertyAccessExpressionWithCapture(context, node, true);
+    }
+    if (ts.isElementAccessExpression(node)) {
+        return transformElementAccessExpressionWithCapture(context, node, true);
+    }
+    return { expression: context.transformExpression(node) };
+}
+
+// similar to moveToPrecedingTemp, except creates an assignment instead of a variable declaration.
+// the assignment is "manually hoisted" in transformOptionalChain so that the variable can be accessed in a later scope.
+export function captureThisValue(
+    context: TransformationContext,
+    expression: lua.Expression,
+    tsOriginal: ts.Node
+): { thisValue: lua.Expression; isNewTemp?: boolean } {
+    if (!shouldMoveToTemp(context, expression, tsOriginal) && !isOptionalContinuation(tsOriginal)) {
+        return { thisValue: expression };
+    }
+    const tempIdentifier = context.createTempNameForLuaExpression(expression);
```

Comment:
I REALLY dislike the fact that tempIdentifier is generated from the expression here and just assigned. This means that this code has a bunch of invisible dependencies on other parts of our code that will be extremely  hard to maintain.

I would propose an alternative: Rename this function to `assignToPrecedingTemp` and pass in the identifier that should be assigned to. You could use this identifier (or undefined) instead of these 'shouldCaptureValue' boolean parameters (call it something like `assignToPrecedingIdentifier?: ts.Identifier`). I think that would make the flow of the code overall much clearer.

Reply:
This would also eliminate the need for these ugly `transformElementAccessExpression() => transformElementAccessExpressionWithCapture` one liners

Reply:
I considered this, but the problem is that there might not be a thisArg to capture, and the temp variable is not always needed.

However, I like your proposal that the temp identifier is passed in; and the temp identifier generated only if necessary. However, info on if the temp was used still needs to be returned, so the variable declaration is only generated/used if needed. I'll implement this in the next commit.

As for the one-liners/extra functions, typescript's own implementation uses an internal AST node called `SyntheticExpression` (`ExpressionWithThisValue` is analgous to this), and has a bunch of assertions that this node doesn't end up outside of `transformES2020`. I opted to go for an explicitly different interface instead, as it would be difficult to ensure an internal node never makes into the result AST, across multiple functions and different files.

Let me know if you would prefer something like typescript's approach (introducing an internal node) or other possible alternatives.